horovod/run/run_func.py

```diff
@@ -0,0 +1,220 @@
+    pickled_proc_fn_path = os.path.join(wdir, _PICKLED_PROC_FN_FILENAME)
+    with open(pickled_proc_fn_path, 'wb') as f:
+        f.write(pickled_proc_fn_str)
+
+    if driver_host is None:
+        driver_host = network.get_local_ip_addr()
+
+    launcher_path = os.path.join(wdir, _PROC_LAUNCHER_FILENAME)
+    with open(launcher_path, 'w') as f:
+        f.write(textwrap.dedent("""
+        set -e
+        cd {wdir}
+        rank=${{OMPI_COMM_WORLD_RANK:-0}}
+        {exec} -c "import cloudpickle; cloudpickle.load(open('{fn_file}', 'rb'))(rank=$rank)"
+        if [[ "$rank" -eq 0 ]]; then
+        scp -q {scp_port_opt} -o StrictHostKeyChecking=no \
+        {wdir}/{local_result} {driver_host}:{wdir}/{result}
+        fi
+        """.format(wdir=wdir,
+                   exec=sys.executable,
+                   scp_port_opt=scp_port_opt,
+                   fn_file=_PICKLED_PROC_FN_FILENAME,
+                   driver_host=driver_host,
+                   local_result=_LOCAL_PICKLED_RESULT_FILENAME,
+                   result=_PICKLED_RESULT_FILENAME)))
+
+    os.chmod(launcher_path, 0o777)
+
+    if host is None and num_proc is None:
+        num_proc = 1
+        host = "localhost:1"
+    elif host is None:
+        host = "localhost:{np}".format(np=num_proc)
+    elif num_proc is None:
+        num_proc = sum(int(x.split(':')[1]) for x in host.split(','))
+
+    all_host_names = list(set(x.split(':')[0] for x in host.split(',')))
+    remote_host_names = network.filter_local_addresses(all_host_names)
+
+    for remote_host in remote_host_names:
+        scp_cmd = textwrap.dedent(
+            """
+            ssh {ssh_port_opt} -o StrictHostKeyChecking=no {remote_host} mkdir -p {wdir_par} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
+            {wdir} {remote_host}:{wdir_par}
+            """).format(ssh_port_opt=ssh_port_opt, scp_port_opt=scp_port_opt, wdir=wdir,
+                        remote_host=remote_host, wdir_par=wdir_par)
+        output = six.StringIO()
+        exit_code = safe_shell_exec.execute(scp_cmd, stdout=output, stderr=output)
+        if exit_code != 0:
+            output_msg = output.getvalue()
+            raise RuntimeError("Copy working dir to remote host {remote_host} failed."
+                               "Error message is {msg}"
+                               .format(remote_host=remote_host, msg=output_msg))
+
+    cmd = ["horovodrun", "-np", str(num_proc)]
+    if ssh_port:
+        cmd += ["-p", str(ssh_port)]
+    if host:
+        cmd += ["-H", str(host)]
+    if disable_cache:
+        cmd += ["--disable-cache"]
+    if start_timeout:
+        cmd += ["--start-timeout", str(start_timeout)]
+    if verbose:
+        cmd += ["--verbose"]
+
+    cmd += ["bash", launcher_path]
+
+    if verbose:
+        print("Run command: " + " ".join(cmd))
+
+    if env is None:
+        env = os.environ
+    task = subprocess.Popen(cmd,
```

Comment:
Should be the same as

~~~python
subprocess.check_call([...], stdin=subprocess.DEVNULL)
~~~